Index: configure
===================================================================
--- configure	(.../vendor/ffmpeg/current)	(revision 775)
+++ configure	(.../moon-ffmpeg)	(revision 775)
@@ -2036,7 +2036,7 @@
 Version: $version
 Requires: $requires
 Conflicts:
-Libs: \${libdir}/${LIBPREF}${shortname}${LIBSUF} $libs
+Libs: \${libdir}/${LIBPREF}${shortname}${SLIBSUF} $libs
 Cflags: -I\${includedir}
 EOF
 }
@@ -2048,16 +2048,16 @@
 pkgconfig_generate_uninstalled libavcodec "FFmpeg codec library" "$lavc_version" "$extralibs" "$pkg_requires libavutil = $lavu_version"
 
 pkgconfig_generate libavformat "FFmpeg container format library" "$lavf_version" "-lavformat $extralibs" "$pkg_requires libavcodec = $lavc_version" ffmpeg
-pkgconfig_generate_uninstalled libavformat "FFmpeg container format library" "$lavf_version" "$extralibs" "$pkg_requires libavcodec = $lavc_version"
+pkgconfig_generate_uninstalled libavformat "FFmpeg container format library" "$lavf_version" "-lavcodec $extralibs" "$pkg_requires libavcodec = $lavc_version"
 
 if enabled pp; then
   pkgconfig_generate libpostproc "FFmpeg post processing library" "$pp_version" -lpostproc "" postproc
-  pkgconfig_generate_uninstalled libpostproc "FFmpeg post processing library" "$pp_version"
+  pkgconfig_generate_uninstalled libpostproc "FFmpeg post processing library" "$pp_version" -lpostproc
 fi
 
 if enabled swscaler; then
   pkgconfig_generate libswscale "FFmpeg image rescaling library" "$sws_version" -lswscale "libavutil = $lavu_version" ffmpeg
-  pkgconfig_generate_uninstalled libswscale "FFmpeg image rescaling library" "$sws_version" "" "libavutil = $lavu_version"
+  pkgconfig_generate_uninstalled libswscale "FFmpeg image rescaling library" "$sws_version" -lswscale "libavutil = $lavu_version"
 else
   pkgconfig_generate libswscale "FFmpeg image rescaling library" "$sws_version" "" "$pkg_requires libavcodec = $lavc_version" ffmpeg
   pkgconfig_generate_uninstalled libswscale "FFmpeg image rescaling library" "$sws_version" "" "$pkg_requires libavcodec = $lavc_version"
Index: libswscale/Makefile
===================================================================
--- libswscale/Makefile	(.../vendor/ffmpeg/current)	(revision 775)
+++ libswscale/Makefile	(.../moon-ffmpeg)	(revision 775)
@@ -5,7 +5,7 @@
 LIBVERSION=$(SWSVERSION)
 LIBMAJOR=$(SWSMAJOR)
 
-EXTRALIBS := -L$(BUILD_ROOT)/libavutil -lavutil$(BUILDSUF) $(EXTRALIBS)
+EXTRALIBS := -L$(BUILD_ROOT)/libavutil $(EXTRALIBS)
 
 OBJS = rgb2rgb.o swscale.o
 
Index: libavcodec/Makefile
===================================================================
--- libavcodec/Makefile	(.../vendor/ffmpeg/current)	(revision 775)
+++ libavcodec/Makefile	(.../moon-ffmpeg)	(revision 775)
@@ -443,7 +443,7 @@
                                           bfin/idct_bfin.o   \
                                           bfin/vp3_idct_bfin.o   \
 
-EXTRALIBS := -L$(BUILD_ROOT)/libavutil -lavutil$(BUILDSUF) $(EXTRALIBS)
+EXTRALIBS := -L$(BUILD_ROOT)/libavutil $(EXTRALIBS)
 
 NAME=avcodec
 LIBVERSION=$(LAVCVERSION)
Index: ChangeLog-moon
===================================================================
--- ChangeLog-moon	(.../vendor/ffmpeg/current)	(revision 0)
+++ ChangeLog-moon	(.../moon-ffmpeg)	(revision 775)
@@ -0,0 +1,13 @@
+2007-10-26  Andrew Jorgensen  <ajorgensen@novell.com>
+
+	* configure: Undo the changes to the .pc files, it's not needed
+	* */Makefile: Don't rename the libraries, this is also not needed
+	because we are dlopening them.  Still we need to change the link flags
+
+2007-10-25  Andrew Jorgensen  <ajorgensen@novell.com>
+
+	* configure: Disable a lot of stuff we don't need (codecs and stuff
+	still need to be disabled) and change the way the .pc files are being
+	generated so that the point to the uninstalled shared objects as I
+	need them to
+	* */Makefile: Rename libraries from libav* to libmoon-av*
Index: libavformat/Makefile
===================================================================
--- libavformat/Makefile	(.../vendor/ffmpeg/current)	(revision 775)
+++ libavformat/Makefile	(.../moon-ffmpeg)	(revision 775)
@@ -162,8 +162,8 @@
 
 OBJS-$(CONFIG_VHOOK)                     += framehook.o
 
-EXTRALIBS := -L$(BUILD_ROOT)/libavutil -lavutil$(BUILDSUF) \
-             -lavcodec$(BUILDSUF) -L$(BUILD_ROOT)/libavcodec $(EXTRALIBS)
+EXTRALIBS := -L$(BUILD_ROOT)/libavutil \
+             -L$(BUILD_ROOT)/libavcodec $(EXTRALIBS)
 
 CPPOBJS-$(CONFIG_AUDIO_BEOS_DEMUXER)     += beosaudio.o
 CPPOBJS-$(CONFIG_AUDIO_BEOS_MUXER)       += beosaudio.o
