Index: text.h
===================================================================
--- text.h	(revision 86192)
+++ text.h	(working copy)
@@ -17,56 +17,8 @@
 #include <pango/pango.h>
 
 #include "brush.h"
-#include "mango.h"
+#include "font.h"
 
-enum FontStretches {
-	FontStretchesUltraCondensed = 1,
-	FontStretchesExtraCondensed = 2,
-	FontStretchesCondensed      = 3,
-	FontStretchesSemiCondensed  = 4,
-	FontStretchesNormal         = 5,
-	FontStretchesMedium         = 5,
-	FontStretchesSemiExpanded   = 6,
-	FontStretchesExpanded       = 7,
-	FontStretchesExtraExpanded  = 8,
-	FontStretchesUltraExpanded  = 9
-};
-
-enum FontStyles {
-	FontStylesNormal,
-	FontStylesOblique,
-	FontStylesItalic
-};
-
-enum FontWeights {
-	FontWeightsThin       = 100,
-	FontWeightsExtraLight = 200,
-	FontWeightsLight      = 300,
-	FontWeightsNormal     = 400,
-	FontWeightsMedium     = 500,
-	FontWeightsSemiBold   = 600,
-	FontWeightsBold       = 700,
-	FontWeightsExtraBold  = 800,
-	FontWeightsBlack      = 900,
-	FontWeightsExtraBlack = 950,
-};
-
-enum StyleSimulations {
-	StyleSimulationsNone
-};
-
-enum TextDecorations {
-	TextDecorationsNone,
-	TextDecorationsUnderline
-};
-
-enum TextWrapping {
-	TextWrappingWrap,
-	TextWrappingNoWrap,
-	TextWrappingWrapWithOverflow
-};
-
-
 void text_init (void);
 void text_destroy (void);
 
@@ -86,7 +38,7 @@
 	virtual void OnPropertyChanged (DependencyProperty *prop);
 	virtual void OnSubPropertyChanged (DependencyProperty *prop, DependencyProperty *subprop);
 
-	PangoFontDescription *font;
+	TextFontDescription *font;
 	Brush *foreground;
 };
 
@@ -137,6 +89,32 @@
 
 
 class TextBlock : public FrameworkElement {
+	TextFontDescription *font;
+	TextLayout *layout;
+	Brush *foreground;
+	
+	bool dirty_actual_values;
+	double actual_height;
+	double actual_width;
+	
+	void CalcActualWidthHeight (cairo_t *cr);
+	void Layout (cairo_t *cr);
+	void Paint (cairo_t *cr);
+	
+	double GetActualWidth ()
+	{
+		if (dirty_actual_values)
+			CalcActualWidthHeight (NULL);
+		return actual_width;
+	}
+	
+	double GetActualHeight ()
+	{
+		if (dirty_actual_values)
+			CalcActualWidthHeight (NULL);
+		return actual_height;
+	}
+	
 public:
 	static DependencyProperty *ActualHeightProperty;
 	static DependencyProperty *ActualWidthProperty;
@@ -171,34 +149,6 @@
 	virtual Value *GetValue (DependencyProperty *property);
 	virtual void SetValue (DependencyProperty *property, Value *value);
 	virtual void SetValue (DependencyProperty *property, Value value);
-	
-private:
-	PangoFontDescription *font;
-	MangoRenderer *renderer;
-	PangoLayout *layout;
-	Brush *foreground;
-	
-	bool dirty_actual_values;
-	double actual_height;
-	double actual_width;
-	
-	void CalcActualWidthHeight (cairo_t *cr);
-	void Layout (cairo_t *cr);
-	void Paint (cairo_t *cr);
-
-	double GetActualWidth ()
-	{
-		if (dirty_actual_values)
-			CalcActualWidthHeight (NULL);
-		return actual_width;
-	}
-
-	double GetActualHeight ()
-	{
-		if (dirty_actual_values)
-			CalcActualWidthHeight (NULL);
-		return actual_height;
-	}
 };
 
 TextBlock *text_block_new (void);
Index: text.cpp
===================================================================
--- text.cpp	(revision 86192)
+++ text.cpp	(working copy)
@@ -22,66 +22,11 @@
 #include "text.h"
 
 
-static PangoStretch
-font_stretch (FontStretches stretch)
-{
-	switch (stretch) {
-	case FontStretchesUltraCondensed:
-		return PANGO_STRETCH_ULTRA_CONDENSED;
-	case FontStretchesExtraCondensed:
-		return PANGO_STRETCH_EXTRA_CONDENSED;
-	case FontStretchesCondensed:
-		return PANGO_STRETCH_CONDENSED;
-	case FontStretchesSemiCondensed:
-		return PANGO_STRETCH_SEMI_CONDENSED;
-	case FontStretchesNormal: // FontStretchesMedium (alias)
-	default:
-		return PANGO_STRETCH_NORMAL;
-	case FontStretchesSemiExpanded:
-		return PANGO_STRETCH_SEMI_EXPANDED;
-	case FontStretchesExpanded:
-		return PANGO_STRETCH_EXPANDED;
-	case FontStretchesExtraExpanded:
-		return PANGO_STRETCH_EXTRA_EXPANDED;
-	case FontStretchesUltraExpanded:
-		return PANGO_STRETCH_ULTRA_EXPANDED;
-	}
-}
+static SolidColorBrush *default_foreground_brush = NULL;
 
-static PangoStyle
-font_style (FontStyles style)
-{
-	switch (style) {
-	case FontStylesNormal:
-	default:
-		return PANGO_STYLE_NORMAL;
-	case FontStylesOblique:
-		return PANGO_STYLE_OBLIQUE;
-	case FontStylesItalic:
-		return PANGO_STYLE_ITALIC;
-	}
-}
-
-static PangoWeight
-font_weight (FontWeights weight)
-{
-	// FontWeights and PangoWeight values map exactly
-	
-	if (weight > 900) {
-		// FontWeighs have values between 100-999, Pango only allows 100-900
-		return (PangoWeight) 900;
-	}
-	
-	return (PangoWeight) weight;
-}
-
-
-SolidColorBrush *default_foreground_brush = NULL;
-	
 static Brush *
 default_foreground (void)
 {
-	
 	if (!default_foreground_brush) {
 		default_foreground_brush = new SolidColorBrush ();
 		Color *color = color_from_str ("black");
@@ -108,12 +53,12 @@
 	foreground = NULL;
 	
 	/* initialize the font description */
-	font = pango_font_description_new ();
+	font = new TextFontDescription ();
 }
 
 Inline::~Inline ()
 {
-	pango_font_description_free (font);
+	delete font;
 	
 	if (foreground != NULL) {
 		foreground->Detach (NULL, this);
@@ -131,19 +76,19 @@
     
 	if (prop == Inline::FontFamilyProperty) {
 		char *family = inline_get_font_family (this);
-		pango_font_description_set_family (font, family);
+		font->SetFamily (family);
 	} else if (prop == Inline::FontSizeProperty) {
 		double size = inline_get_font_size (this);
-		pango_font_description_set_absolute_size (font, size * PANGO_SCALE);
+		font->SetSize (size);
 	} else if (prop == Inline::FontStretchProperty) {
 		FontStretches stretch = inline_get_font_stretch (this);
-		pango_font_description_set_stretch (font, font_stretch (stretch));
+		font->SetStretch (stretch);
 	} else if (prop == Inline::FontStyleProperty) {
 		FontStyles style = inline_get_font_style (this);
-		pango_font_description_set_style (font, font_style (style));
+		font->SetStyle (style);
 	} else if (prop == Inline::FontWeightProperty) {
 		FontWeights weight = inline_get_font_weight (this);
-		pango_font_description_set_weight (font, font_weight (weight));
+		font->SetWeight (weight);
 	} else if (prop == Inline::ForegroundProperty) {
 		if (foreground != NULL) {
 			foreground->Detach (NULL, this);
@@ -321,8 +266,6 @@
 
 TextBlock::TextBlock ()
 {
-	renderer = (MangoRenderer *) mango_renderer_new ();
-	
 	foreground = NULL;
 	
 	layout = NULL;
@@ -332,17 +275,17 @@
 	dirty_actual_values = true;
 	
 	/* initialize the font description */
-	font = pango_font_description_new ();
+	font = new TextFontDescription ();
 	char *family = text_block_get_font_family (this);
-	pango_font_description_set_family (font, family);
+	font->SetFamily (family);
 	double size = text_block_get_font_size (this);
-	pango_font_description_set_absolute_size (font, size * PANGO_SCALE);
+	font->SetSize (size);
 	FontStretches stretch = text_block_get_font_stretch (this);
-	pango_font_description_set_stretch (font, font_stretch (stretch));
+	font->SetStretch (stretch);
 	FontStyles style = text_block_get_font_style (this);
-	pango_font_description_set_style (font, font_style (style));
+	font->SetStyle (style);
 	FontWeights weight = text_block_get_font_weight (this);
-	pango_font_description_set_weight (font, font_weight (weight));
+	font->SetWeight (weight);
 	
 	// this has to come last, since in our OnPropertyChanged
 	// method we update our bounds.
@@ -352,13 +295,11 @@
 
 TextBlock::~TextBlock ()
 {
-	pango_font_description_free (font);
+	delete font;
 	
 	if (layout)
-		g_object_unref (layout);
+		delete layout;
 	
-	g_object_unref (renderer);
-	
 	if (foreground != NULL) {
 		foreground->Detach (NULL, this);
 		foreground->unref ();
@@ -374,11 +315,6 @@
 void
 TextBlock::Render (cairo_t *cr, int x, int y, int width, int height)
 {
-	const char *text = pango_layout_get_text (layout);
-	
-	if (!text || !text[0])
-		return;
-	
 	cairo_save (cr);
 	cairo_set_matrix (cr, &absolute_xform);
 	Paint (cr);
@@ -434,17 +370,12 @@
 void
 TextBlock::Layout (cairo_t *cr)
 {
-	PangoAttribute *uline_attr = NULL;
-	PangoAttribute *font_attr = NULL;
-	PangoAttribute *fg_attr = NULL;
-	PangoAttribute *attr = NULL;
 	TextDecorations decorations;
 	double clip_height, width;
-	PangoFontMask font_mask;
-	PangoAttrList *attrs;
-	size_t start, end;
+	TextWrapping wrapping;
+	uint8_t font_mask;
 	bool clip = false;
-	GString *block;
+	List *runs;
 	char *text;
 	Brush *fg;
 	int w, h;
@@ -455,68 +386,40 @@
 		fg = foreground;
 	
 	if (layout == NULL)
-		layout = pango_cairo_create_layout (cr);
+		layout = new TextLayout ();
 	
 	clip_height = framework_element_get_height (this);
-	switch (text_block_get_text_wrapping (this)) {
+	wrapping = text_block_get_text_wrapping (this);
+	layout->SetWrapping (wrapping);
+	switch (wrapping) {
 	case TextWrappingWrap:
 		// same as w/ Overflow except we clip height (if defined)
 		if (clip_height > 0.0)
 			clip = true;
 	case TextWrappingWrapWithOverflow:
-		pango_layout_set_wrap (layout, PANGO_WRAP_WORD_CHAR);
+		if ((width = framework_element_get_width (this)) <= 0.0f)
+			width = -1.0f;
 		
-		width = framework_element_get_width (this);
-		
-		if (width > 0.0)
-			pango_layout_set_width (layout, (int) width * PANGO_SCALE);
-		else
-			pango_layout_set_width (layout, -1);
+		layout->SetMaxWidth ((int) width);
 		break;
 	default:
-		pango_layout_set_width (layout, -1);
+		layout->SetMaxWidth (-1);
 		break;
 	}
 	
-	block = g_string_new ("");
-	attrs = pango_attr_list_new ();
+	runs = new List ();
 	
-	font_mask = pango_font_description_get_set_fields (font);
+	font_mask = font->GetFields ();
 	decorations = text_block_get_text_decorations (this);
 	text = text_block_get_text (this);
-	if (text && *text) {
-		g_string_append (block, text);
-		end = block->len;
-		start = 0;
-		
-		font_attr = pango_attr_font_desc_new (font);
-		font_attr->start_index = start;
-		font_attr->end_index = end;
-		
-		pango_attr_list_insert (attrs, font_attr);
-		
-		if (decorations == TextDecorationsUnderline) {
-			uline_attr = pango_attr_underline_new (PANGO_UNDERLINE_SINGLE);
-			uline_attr->start_index = start;
-			uline_attr->end_index = end;
-			
-			pango_attr_list_insert (attrs, uline_attr);
-		} else {
-			uline_attr = NULL;
-		}
-		
-		fg_attr = mango_attr_foreground_new (this, fg);
-		fg_attr->start_index = start;
-		fg_attr->end_index = end;
-		
-		pango_attr_list_insert (attrs, fg_attr);
-	}
+	if (text && *text)
+		runs->Append (new TextRun (text, -1, decorations, font, fg));
 	
 	Inlines *inlines = text_block_get_inlines (this);
 	
 	if (inlines != NULL) {
 		Collection::Node *node = (Collection::Node *) inlines->list->First ();
-		PangoFontMask run_mask, inherited_mask;
+		uint8_t run_mask, inherited_mask;
 		TextDecorations deco;
 		Value *value;
 		Inline *item;
@@ -525,103 +428,53 @@
 		while (node != NULL) {
 			item = (Inline *) node->obj;
 			
+			// Inlines inherit their parent TextBlock's font properties if
+			// they don't specify their own.
+			run_mask = item->font->GetFields ();
+			item->font->Merge (font, false);
+			
+			inherited_mask = (FontMask) (font_mask & ~run_mask);
+			
+			// Inherit the TextDecorations from the parent TextBlock if unset
+			value = item->GetValue (Inline::TextDecorationsProperty);
+			deco = value ? (TextDecorations) value->AsInt32 () : decorations;
+			
 			switch (item->GetObjectType ()) {
 			case Type::RUN:
 				run = (Run *) item;
 				
 				text = run_get_text (run);
 				
-				if (text == NULL || *text == '\0') {
-					// optimization
-					goto loop;
+				if (text && text[0]) {
+					if (item->foreground)
+						runs->Append (new TextRun (text, -1, deco, item->font, item->foreground));
+					else
+						runs->Append (new TextRun (text, -1, deco, item->font, fg));
 				}
 				
-				start = block->len;
-				g_string_append (block, text);
-				end = block->len;
 				break;
 			case Type::LINEBREAK:
-				start = block->len;
-				g_string_append_c (block, '\n');
-				end = block->len;
+				if (item->foreground)
+					runs->Append (new TextRun (deco, item->font, item->foreground));
+				else
+					runs->Append (new TextRun (deco, item->font, fg));
 				break;
 			default:
-				goto loop;
 				break;
 			}
 			
-			// Inlines inherit their parent TextBlock's font properties if
-			// they don't specify their own.
-			run_mask = pango_font_description_get_set_fields (item->font);
-			pango_font_description_merge (item->font, font, false);
-			inherited_mask = (PangoFontMask) (font_mask & ~run_mask);
-			
-			attr = pango_attr_font_desc_new (item->font);
-			attr->start_index = start;
-			attr->end_index = end;
-			
-			if (!font_attr || !pango_attribute_equal ((const PangoAttribute *) font_attr, (const PangoAttribute *) attr)) {
-				pango_attr_list_insert (attrs, attr);
-				font_attr = attr;
-			} else {
-				pango_attribute_destroy (attr);
-				font_attr->end_index = end;
-			}
-			
 			if (inherited_mask != 0)
-				pango_font_description_unset_fields (item->font, inherited_mask);
+				item->font->UnsetFields (inherited_mask);
 			
-			// Inherit the TextDecorations from the parent TextBlock if unset
-			value = item->GetValue (Inline::TextDecorationsProperty);
-			deco = value ? (TextDecorations) value->AsInt32 () : decorations;
-			if (deco == TextDecorationsUnderline) {
-				if (uline_attr == NULL) {
-					uline_attr = pango_attr_underline_new (PANGO_UNDERLINE_SINGLE);
-					uline_attr->start_index = start;
-					uline_attr->end_index = end;
-					
-					pango_attr_list_insert (attrs, uline_attr);
-				} else {
-					uline_attr->end_index = end;
-				}
-			} else {
-				uline_attr = NULL;
-			}
-			
-			// Inlines also inherit their Foreground property from their parent
-			// TextBlock if not set explicitly
-			if (item->foreground)
-				attr = mango_attr_foreground_new (this, item->foreground);
-			else
-				attr = mango_attr_foreground_new (this, fg);
-			attr->start_index = start;
-			attr->end_index = end;
-			
-			if (!fg_attr || !pango_attribute_equal ((const PangoAttribute *) fg_attr, (const PangoAttribute *) attr)) {
-				pango_attr_list_insert (attrs, attr);
-				fg_attr = attr;
-			} else {
-				pango_attribute_destroy (attr);
-				fg_attr->end_index = end;
-			}
-			
-		loop:
 			node = (Collection::Node *) node->next;
 		}
 	}
 	
-	// Now that we have our PangoAttrList setup, set it and the text on the PangoLayout
-	pango_layout_set_text (layout, block->str, block->len);
-	g_string_free (block, true);
+	layout->SetTextRuns (runs);
+	layout->Layout ();
 	
-	pango_layout_set_attributes (layout, attrs);
+	layout->GetPixelSize (&w, &h);
 	
-	pango_cairo_update_layout (cr, layout);
-	mango_renderer_set_cairo_context (renderer, cr);
-	mango_renderer_layout_path (renderer, layout);
-	pango_layout_get_pixel_size (layout, &w, &h);
-	pango_attr_list_unref (attrs);
-	
 	if (clip && (h > clip_height))
 		text_block_set_actual_height (this, (double) clip_height);
 	else
@@ -636,7 +489,7 @@
 	TextWrapping wrapping = text_block_get_text_wrapping (this);
 	double h = GetActualHeight ();
 	double w = framework_element_get_width (this);
-
+	
 	if (wrapping != TextWrappingWrapWithOverflow)
 		h = framework_element_get_height (this);
 	
@@ -645,9 +498,7 @@
 		cairo_clip (cr);
 	}
 	
-	pango_cairo_update_layout (cr, layout);
-	mango_renderer_set_cairo_context (renderer, cr);
-	mango_renderer_show_layout (renderer, layout);
+	layout->Render (cr, this, 0.0, 0.0);
 }
 
 void
@@ -663,19 +514,19 @@
 	
 	if (prop == TextBlock::FontFamilyProperty) {
 		char *family = text_block_get_font_family (this);
-		pango_font_description_set_family (font, family);
+		font->SetFamily (family);
 	} else if (prop == TextBlock::FontSizeProperty) {
 		double size = text_block_get_font_size (this);
-		pango_font_description_set_absolute_size (font, size * PANGO_SCALE);
+		font->SetSize (size);
 	} else if (prop == TextBlock::FontStretchProperty) {
 		FontStretches stretch = text_block_get_font_stretch (this);
-		pango_font_description_set_stretch (font, font_stretch (stretch));
+		font->SetStretch (stretch);
 	} else if (prop == TextBlock::FontStyleProperty) {
 		FontStyles style = text_block_get_font_style (this);
-		pango_font_description_set_style (font, font_style (style));
+		font->SetStyle (style);
 	} else if (prop == TextBlock::FontWeightProperty) {
 		FontWeights weight = text_block_get_font_weight (this);
-		pango_font_description_set_weight (font, font_weight (weight));
+		font->SetWeight (weight);
 	} else if (prop == TextBlock::TextProperty) {
 		// will be updated later in Layout()
 	} else if (prop == TextBlock::InlinesProperty) {
@@ -1126,6 +977,8 @@
 void
 text_init (void)
 {
+	font_init ();
+	
 	// Inline
 	Inline::FontFamilyProperty = DependencyObject::Register (Type::INLINE, "FontFamily", Type::STRING);
 	Inline::FontSizeProperty = DependencyObject::Register (Type::INLINE, "FontSize", Type::DOUBLE);
